### PR TITLE
MINOR : KIP-297 : use service loader for ConfigProvider impls

### PR DESCRIPTION
**Pull request by rayokota**
This is a small change to use the Java ServiceLoader to load ConfigProvider plugins.  It uses code added by @mageshn for Connect Rest Extensions.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: clients/src/main/java/org/apache/kafka/common/config/FileConfigProvider.java [hunk 1]
```diff
@@ -16,6 +16,7 @@
  */
 package org.apache.kafka.common.config;
 
+import java.io.BufferedReader;
```
Comment (mageshn) [inline clients/src/main/java/org/apache/kafka/common/config/FileConfigProvider.java:19]: Should FileConfigProvider class be listed as whitelist in PluginUtils?

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerConfig.java [hunk 2]
```diff
@@ -192,16 +192,20 @@
...
 
     public static final String REST_EXTENSION_CLASSES_CONFIG = "rest.extension.classes";
     protected static final String REST_EXTENSION_CLASSES_DOC =
             "Comma-separated names of <code>ConnectRestExtension</code> classes, loaded and called "
             + "in the order specified. Implementing the interface  "
-            + "<code>ConnectRestExtension</code> allows you to inject into Connect's REST API user defined resources  like filters. "
-            + "Typically used to add custom capability like logging, security, etc.";
+            + "<code>ConnectRestExtension</code> allows you to inject into Connect's REST API user defined resources like filters. "
+            + "Typically used to add custom capability like logging, security, etc. "
+            + "The implementations will be discovered using the standard Java <code>ServiceLoader</code> mechanism.";
```
Comment (mageshn) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerConfig.java]: Like this 👍

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerConfig.java [hunk 3]
```diff
@@ -192,16 +192,20 @@
...
     public static final String CONFIG_PROVIDERS_CONFIG = "config.providers";
-    protected static final String CONFIG_PROVIDERS_DOC = "List of configuration providers. "
-            + "This is a comma-separated list of the fully-qualified names of the ConfigProvider implementations, "
-            + "in the order they will be created, configured, and used.";
+    protected static final String CONFIG_PROVIDERS_DOC =
+            "Comma-separated names of <code>ConfigProvider</code> classes, loaded and used "
+            + "in the order specified. Implementing the interface  "
+            + "<code>ConfigProvider</code> allows you to replace variable references in connector configurations, "
+            + "such as for externalized secrets. "
+            + "The implementations will be discovered using the standard Java <code>ServiceLoader</code> mechanism.";
```
Comment (rhauch) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerConfig.java]: This last sentence is probably not for the user audience that will simply try to configure a worker. I suggest removing it.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerConfig.java (hunk 2 above)
Comment (rhauch) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerConfig.java]: I suggest removing this last sentence, too, for the same reason.

**Review by rhauch — COMMENTED** [review]
Two minor suggestions. Otherwise looks good.

**Review by mageshn — COMMENTED** [review]
Thanks, @rayokota. Just one minor comment. Also, have you considered using <T> T newPlugin(String klassName, AbstractConfig config, Class<T> pluginKlass) in lieu of newConfigProvider? Maybe the only change that is needed is if you would want to be able to configure just configs with a prefix.

File: clients/src/main/java/org/apache/kafka/common/config/FileConfigProvider.java (hunk 1 above)
Comment (mageshn) [reply clients/src/main/java/org/apache/kafka/common/config/FileConfigProvider.java:19]: Also, maybe we need a Manifest for FileConfigProvider

**Comment by rayokota** [comment]
@rhauch , I removed the comments as you suggested.  @mageshn , I added a manifest to the Kafka client jar.  The `newConfigProvider` does a bit more than `newPlugin` so I left it for now.

**Comment by mageshn** [comment]
Thanks, @rayokota how about listing  the FileConfigProvider class as whitelist in PluginUtils

**Comment by rayokota** [comment]
@mageshn , I added a pattern for ConfigProviders to the WHITE_LIST

**Review by mageshn — APPROVED** [review]
LGTM. Thanks, @rayokota.

**Comment by rayokota** [comment]
Hi @ewencp , if you get a chance, do you think you can take a look and merge if this looks ok?

**Review by rhauch — APPROVED** [review]
LGTM.

**Comment by ewencp** [comment]
LGTM, merging to trunk and 2.0

**Comment by rayokota** [comment]
Thanks @ewencp !

**Pull request closed without merging**